libs/wire-api/test/unit/Test/Wire/API/Golden/Runner.hs

```diff
@@ -34,37 +35,87 @@ import Data.ProtoLens.Message (Message)
 import Data.ProtoLens.TextFormat (pprintMessage, readMessage)
 import qualified Data.Text.Lazy.IO as LText
 import Imports
+import Polysemy (Embed, Members, Sem, embed, runM)
+import Polysemy.Error (Error, runError, throw)
 import Test.Tasty.HUnit
 import Text.PrettyPrint (render)
 import Type.Reflection (typeRep)
 import Wire.API.ServantProto
 
+data TestOutcomeError a
+  = FileDoesNotExist FilePath
+  | FailedToDecode String
+  | AesonValueMismatch
+      Value
+      -- ^ Expected value
+      Value
+      -- ^ Actual value
+      FilePath
+      -- ^ The path to the file
+  | FailedToParse a (Result a) FilePath
+
+instance (Typeable a, Show a) => Show (TestOutcomeError a) where
+  show = \case
+    FileDoesNotExist path -> "Golden JSON file " <> show path <> " does not exist"
+    FailedToDecode err -> "Failed to decode: " <> err
+    AesonValueMismatch expected actual path ->
+      show (typeRep @a) <> ": ToJSON should match golden file: " <> show path
+        <> "\n"
+        <> "expected: "
+        <> show expected
+        <> "\n"
+        <> " but got: "
+        <> show actual
+    FailedToParse obj result path ->
+      (show (typeRep @a) <> ": FromJSON of " <> show path <> " should match object")
+        <> "\n"
+        <> "expected: "
+        <> show (Success obj)
+        <> "\n"
+        <> " but got: "
+        <> show result
+
 testObjects :: forall a. (Typeable a, ToJSON a, FromJSON a, Eq a, Show a) => [(a, FilePath)] -> IO ()
 testObjects objs = do
-  allFilesExist <- and <$> traverse (uncurry testObject) objs
-  assertBool "Some golden JSON files do not exist" allFilesExist
+  allResults <- traverse (run . uncurry testObject) objs
+  let errorResults = foldMap (either pure (const [])) allResults
+      errorMsgs = intercalate "\n" . fmap show $ errorResults
+  assertBool errorMsgs . null $ errorResults
+  where
+    run :: Sem '[Error (TestOutcomeError a), Embed IO] () -> IO (Either (TestOutcomeError a) ())
+    run = runM . runError
 
-testObject :: forall a. (Typeable a, ToJSON a, FromJSON a, Eq a, Show a) => a -> FilePath -> IO Bool
+-- | A passing test returns a '()'. If there is a failure, a 'TestOutcomeError
+-- a' is thrown in the 'Sem r' monad.
+testObject ::
+  forall a r.
+  ( Members '[Error (TestOutcomeError a), Embed IO] r,
+    Typeable a,
+    ToJSON a,
+    FromJSON a,
+    Eq a,
+    Show a
+  ) =>
+  a ->
+  FilePath ->
+  Sem r ()
```

Comment:
I think it would be much simpler if this function just returned `IO (Either TestOutcomeError ())` instead of involving polysemy here.

Reply:
That's a perfectly valid choice. In that case we don't have a short-circuiting mechanism of a monad with errors (like we do with `Sem r` that has `Error`s) and instead we have to implement our own nested if-then-else thingy. Should I do that?

Reply:
We are pretty much doing the nested conditional logic anyway. But I think it would be more promising if we relied on tasty to do all of this for us by using `testCase` for each test object instead of making all these assertions in one `testCase`.

Reply:
I just got rid of Polysemy here, and each test object is now a test case on its own.